flow/record/adapter/splunk.py

```diff
@@ -1,88 +1,280 @@
+    HTTP = "http"
+    HTTPS = "https"
+    TCP = "tcp"
+
+
+class SourceType(Enum):
+    JSON = "json"
+    RECORDS = "records"
+
+
+def splunkify_key_value(record: Record, tag: Optional[str] = None) -> str:
     ret = []
 
-    ret.append(f'type="{record._desc.name}"')
+    ret.append(f'rdtype="{record._desc.name}"')
 
     if tag is None:
         ret.append("rdtag=None")
     else:
         ret.append(f'rdtag="{tag}"')
 
-    for field in record._desc.fields:
+    for field in record._desc.get_all_fields():
+        # Omit the _version field as the Splunk adapter has no reader support for deserialising records back.
+        if field == "_version":
+            continue
+
         val = getattr(record, field)
+
+        if field in PREFIX_WITH_RD:
+            field = f"rd_{field}"
+
         if val is None:
             ret.append(f"{field}=None")
         else:
             val = to_base64(val) if isinstance(val, bytes) else to_str(val)
             val = val.replace("\\", "\\\\").replace('"', '\\"')
-            if field in RESERVED_SPLUNK_FIELDS:
-                field = f"rd_{field}"
             ret.append(f'{field}="{val}"')
 
     return " ".join(ret)
 
 
+def splunkify_json(packer: JsonRecordPacker, record: Record, tag: Optional[str] = None) -> str:
+    ret = {}
+
+    indexer_fields = [
+        ("host", "host"),
+        ("host", "hostname"),
+        ("time", "ts"),
+    ]
+
+    # When converting a record to json text for splunk, we distinguish between the 'event' (containing the data) and a
+    # few other fields that are splunk-specific for indexing. We add those 'indexer_fields' to the return object first.
+    for splunk_name, field_name in indexer_fields:
+        if hasattr(record, field_name):
+            val = getattr(record, field_name)
+            if val:
+                if isinstance(val, datetime):
+                    # Convert datetime objects to epoch timestamp for reserved fields.
+                    ret[splunk_name] = val.timestamp()
+                    continue
+                ret[splunk_name] = to_str(val)
+
+    record_as_dict = packer.pack_obj(record)
+
+    # Omit the _version field as the Splunk adapter has no reader support for deserialising records back.
+    del record_as_dict["_version"]
+
+    # These fields end up in the 'event', but we have a few reserved field names. If those field names are in the
+    # record, we prefix them with 'rd_' (short for record descriptor)
+    for field in PREFIX_WITH_RD:
+        if field not in record_as_dict:
+            continue
+        new_field = f"rd_{field}"
+
+        record_as_dict[new_field] = record_as_dict[field]
+        del record_as_dict[field]
+
+    # Almost done, just have to add the tag and the type (i.e the record descriptor's name) to the event.
+    record_as_dict["rdtag"] = tag
+
+    # Yes.
+    record_as_dict["rdtype"] = record._desc.name
+
+    ret["event"] = record_as_dict
+    return json.dumps(ret, default=packer.pack_obj)
```

Comment:
Turns out that the `JsonRecordPacker` leaves record values of type `posix_path` as-is. Because of that, `record_as_dict` will contain non-serializable fields. When `json.dumps` attempts to serialize a value of type `posix_path`, it will cal the `JsonRecordPacker` on that individual value, whereas earlier it was called on _a record containing said value_. `JsonRecordPacker` will then convert it to a string.

Meaning, leaving default out here will break on records containing `posix_path` field types. We could look into fixing this in the `JsonRecordPacker`, or we could drop everything and go home, seeing as it's 5PM and it's a Friday.

Reply:
Ah yeah that looks like an oversight when adding the `path` type, that should've been added to the JSON packer as well. Let's fix that in another PR.

See you at home.

Reply:
> Ah yeah that looks like an oversight when adding the `path` type, that should've been added to the JSON packer as well. Let's fix that in another PR.
> 
> See you at home.

Was this eventually fixed in another PR? If not, please make an issue.